### ISSUE TITLE
What's the purpose of defaults/main directory?

### ISSUE THREAD

**Issue by bschonec** [Thulium-Drake/ansible-role-vmware#1]
Not so much an issue but a use problem.  I understand the defaults/main.yml for role variables but what is the purpose of the main *directory*?  Does Ansible automatically load the variables for all YAML files located under the defaults/main directory?

**Comment by Thulium-Drake** [comment]
Hey! Thanks for the feedback!

Yes, that is exactly what happens, I did that as to not clutter the variables files too much, so it's more clear which defaults belong with which playbook.

**Comment by bschonec** [comment]
Ah, makes sense now.  Mind if I update the README.md for those semi-noobs like me who might know what the default behavior is?

I've not found anything (yet) on the ansible documentation site that describes that behavior.

**Comment by Thulium-Drake** [comment]
Sure thing! I'll see the PR :-)

Because this role is new and it's the first one I actually published, can you also let me know how you use it and if there are any quirks in the usage I might have missed? I've tried to make it as easy as possible :-)

**Comment by bschonec** [comment]
To be honest, I have been trying for about eight hours to get the role to work.  I don't quite understand the 'hosts: all' and then the variable, 'target_group'.  From what I'm [mis]understanding, wouldn't I want to specify the group of hosts in "hosts: <group>" instead?  The "hosts: all" makes me a bit nervous.  ;-)

I'm testing on Red Hat and would like to be able to skip the vmware_install.yml task (I'll put in a PR) since Python's PIP command can do some serious damage the the Red Hat supplied Python RPMs.

I don't understand why the individual hosts would need the following: https://github.com/Thulium-Drake/ansible-role-vmware/blob/1f9a6840f00ea9ebf5801432c04d471faa4fe404/tasks/main.yml#L6

I'm not sure why the individual hosts would need to do anything "vmware-ish" since all they need to do is determine their name (fqdn?) and/or VMWare UUID via 'dmidecode.'  Could you elaborate a bit on that?

In the very simple role that I created before discovering yours, the vsphere_api_host machine does all the heavy lifting by talking the the vcenter cluster(s).

I can't get past: https://github.com/Thulium-Drake/ansible-role-vmware/blob/1f9a6840f00ea9ebf5801432c04d471faa4fe404/tasks/vmware_install.yml#L9 because it errors out with: 

    fatal: [brian-lab]: FAILED! => {"msg": "'ansible.vars.hostvars.HostVarsVars object' has no attribute 'vsphere_api_host'"}



...

**Comment by Thulium-Drake** [comment]
Ah! It seems that you missed the delegation part, I've made the role do the following:

* You configure all the variables for VMWare on groups and/or hosts
* When this role is run, Ansible will lookup the variables for the current target host
* It then connects to the specified (in the vars) delegate host where the VMWare API SDK has been installed
* The delegated host will contact VMWare to do the requested tasks

Check here for an example setup: https://github.com/Thulium-Drake/ansible-examples/tree/master/vmware-example-setup

That will also tell you a bit more about the RHEL setup. By default, the RHEL repo's do not feature pip and the only way to install it is by using RHSCL, which does offer pip. When you install it, it will create a virtualenv-like environment in /opt/rh/python2.7 (or something like that).

This role will set this up and then error out at the end of the installation playbook:
https://github.com/Thulium-Drake/ansible-role-vmware/blob/1f9a6840f00ea9ebf5801432c04d471faa4fe404/tasks/vmware_install.yml#L70-L85

So installing it this way doesn't do any harm to any existing Python libs that are installed on the system. In fact, when you run the role again, it will attempt the installation process again, because the system python environment (the one you get when running the default python interpreter, like Ansible does) does not contain the modules you need to talk to VMWare.

Does this clarify it a bit more?

EDIT: I have also clarified the example playbooks a bit more

**Comment by bschonec** [comment]
I'm really confused as to where "vsphere_api_host" comes from.   I have set the variable in my playbook when invoking your roll but the managed server still doesn't seem to have that property.

I've put it in my playbook but I still get:

fatal: [brian-lab]: FAILED! => {"msg": "'ansible.vars.hostvars.HostVarsVars object' has no attribute 'vsphere_api_host'"}

Is there some soft of VMWare python module that adds facts to the vsphere_api_host?  Even debugging the output of the entire {{ hostvars[<item>] }}  hash shows no "api" type variables/facts.

I've verified that the VMWare python dev kit is working because I have a few playbooks that do snapshots, power resets, etc.  I'm just unable to get your role working....admittedly you're doing some fanciness that I don't yet understand.

**Comment by Thulium-Drake** [comment]
The 'vsphere_api_host' is a variable that points to the host where you have installed the VMWare SDK. Ansible will contact that machine for all the API calls towards VMWare.

The whole idea behind the role is that you can specify all variables per host/hostgroup.

So, imagine you have 1 VMWare cluster for all VMs, you have all these vars in group_vars/all.yml (copy them from the defaults folder)
```
---
# group_vars/all.yml
# The host that has pyVmomi installed
vsphere_api_host: 'ansible.example.com'
# Your vCenter machine
vsphere_host: 'vcenter-prd.example.com'
# Whether to check vSphere's SSL cert or not
vsphere_validate_certs: 'no'
# A useraccount with sufficient privileges
vsphere_username: 'administrator@vsphere.local'
# The password for the above user
vsphere_password: 'Password1!'
# Which Datacenter your VM is in
vsphere_datacenter: 'prod'
```

If you have a secondary cluster for all test machines, which are member of the 'test' group, you create an additional vars file in group_vars/test.yml and fill in the information for them. But you don't have to fill in all the vars that are the same, for example, you only connect to a different vSphere server, but the useraccount is the same:

```
---
# group_vars/test.yml
# Your vCenter machine
vsphere_host: 'vcenter-test.example.com'
# Which Datacenter your VM is in
vsphere_datacenter: 'test'
```

The above setup allows for setting global defaults and make overrides for specific groups.

You could even make vars file for a single host, if that host happens to be on yet another vSphere environment.

**Comment by bschonec** [comment]
I must be missing something fundamental.  I am familiar with roles and your role is loading just fine.  It's the vsphere_api_host variable that's not getting 'loaded' into my servers.

My playbook that calls your role:

https://gist.github.com/bschonec/c67779f2a62674c4ee7b6eedf9b88b01

My inventory file:

https://gist.github.com/bschonec/c67779f2a62674c4ee7b6eedf9b88b01

Results:

https://gist.github.com/bschonec/72504e13212a1853f4654b2354a9d0a3


"brian-lab" is the name of the server that talks to Vmware.  "gitlab-all" is a one-member group that contains the name of the server I want to make a snap of.  The role fails on the main.yml task in the "- name: 'Check dependencies' play.

My apologies for all the trouble...

**Comment by Thulium-Drake** [comment]
No worries ;-)

But I see the problem! You didn't configure them as host variables, but as variables on the role. Try putting them in the group_vars.

This might need some additional documentation to clarify the differences. It boils down to:

All values that are found in the 'defaults' folder are HOST variables (and need to be configured as such.

All the values that are found in the 'vars' folder are ROLE variables, which are just the following:

* target_action
* target_state
* target_group
* target_snapshot_name

I'm thinking out loud now, this might be fixed by making a clear naming scheme for all variables. Would the following scheme make things clear for you:

* target_ / vsphere_ vars are about hosts, therefore need to be defined as host variables
* input_ vars are for steering the role

This will require some refactoring of the variables, but if it makes it easier to work with, I'm fine with that.

**Comment by bschonec** [comment]
Woo!  I'm closer.  I'm able to get past the 

TASK [vmware : Check dependencies] ************************************************************************************************************************************

section but it's not yet connecting to the vmware server.  I'll investigate further.

I did notice that I get different behavior when the variables are in ../group_vars/all.yml versus the group variables in the inventory file as described by:  [this ansible doc](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html#group-variables)

Do you have a recommendation/preference?

**Comment by Thulium-Drake** [comment]
Are you running a RHEL based machine? Then you need to do the following:

1. Run the role against your RHEL machine, it will then install the RHSCL environment and install the required python modules in it. It will also place 2 scripts in /usr/local/bin.
2. Change the python interpreter for your RHEL machine to /usr/local/bin/ansible_vmware_python
3. Run the role again.

If you skip 1, it will always check the installation and fail at the end.

> Do you have a recommendation/preference?

I've never really defined vars inside the inventory, until now I've almost always used group_vars/$group_name.yml

**Comment by Thulium-Drake** [comment]
I have done some testing, I think I may have made the lookup part too complicated. Give me 10 min to push a new version.

EDIT: There's some stuff in a new branch to try, but I'm still working on it :-)

EDIT2: Nope, this actually breaks it :-( this will break host/group-based exceptions for which vCenter server to use.

So resuming, some variables can be made host specific, grepping on the code gives me this list:

vsphere_api_host
vsphere_datacenter
vsphere_host
vsphere_password
vsphere_username

So, these _have_ to be defined on the host.

EDIT3: I have commited a change that removes the target_group variable, it doesn't really add anything, I have changed all loops to loop over all hosts currently targeted by the play (which I basically already did by assigning ansible_play_hosts to target_group)
